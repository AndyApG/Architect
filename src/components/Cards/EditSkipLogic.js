import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, compose } from 'redux';
import PropTypes from 'prop-types';
import { get, find } from 'lodash';
import { Button } from '../../ui/components';
import { actionCreators as stageActions } from '../../ducks/modules/protocol/stages';
import Card from './ProtocolCard';
import { getProtocol } from '../../selectors/protocol';
import { Draft } from '../../behaviours';
import SkipLogicEditor from '../SkipLogicEditor';

const defaultLogic = {
  action: 'SKIP',
  operator: '',
  value: '',
  filter: {
    join: '',
    rules: [],
  },
};

class EditSkipLogic extends PureComponent {
  static propTypes = {
    show: PropTypes.bool,
    hasChanges: PropTypes.bool,
    stageId: PropTypes.string,
    onComplete: PropTypes.func,
    draft: PropTypes.any.isRequired,
    updateStage: PropTypes.func.isRequired,
    updateDraft: PropTypes.func.isRequired,
  };

  static defaultProps = {
    show: false,
    stageId: null,
    hasChanges: false,
    onComplete: () => {},
  }

  onSave = () => {
    const stageId = this.props.stageId;

    this.props.updateStage(
      stageId,
      {
        skipLogic: this.props.draft,
      },
    );

    this.props.onComplete();
  };

  renderButtons() {
    return [].concat(
      this.props.hasChanges ? [<Button key="save" size="small" onClick={this.onSave}>Continue</Button>] : [],
    );
  }

  render() {
    const {
      show,
      onComplete,
      draft,
      updateDraft,
    } = this.props;

    return (
      <Card
        type="logic"
        buttons={this.renderButtons()}
        show={show}
        onCancel={onComplete}
      >
        <SkipLogicEditor onChange={updateDraft} rules={draft} />
      </Card>
    );
  }
}

const mapStateToProps = (state, props) => {
  const stageId = get(props, 'match.params.id');
  const protocol = getProtocol(state);
  const stage = find(protocol.stages, ['id', stageId]);
  const logic = get(stage, 'skipLogic', defaultLogic);

  return {
    stageId,
    draft: logic,
  };
};

const mapDispatchToProps = dispatch => ({
  updateStage: bindActionCreators(stageActions.updateStage, dispatch),
});

export { EditSkipLogic };

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  Draft,
)(EditSkipLogic);
