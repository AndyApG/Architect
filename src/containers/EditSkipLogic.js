import React, { PureComponent } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, compose } from 'redux';
import PropTypes from 'prop-types';
import { has, toPairs } from 'lodash';
import { Button } from 'network-canvas-ui';
import { makeGetStage } from '../selectors/protocol';
import { actionCreators as stageActions } from '../ducks/modules/protocol/stages';
import { ProtocolCard, FilterGroup } from '../containers';
import { NetworkRule } from '../containers/Rule';
import { Draft } from '../behaviours';
import DropDown from '../components/Rule/DropDown';

const defaultLogic = {
  action: 'SKIP',
  operator: '',
  value: '',
  filter: {
    join: '',
    rules: [],
  },
};

class EditSkipLogic extends PureComponent {
  static propTypes = {
    show: PropTypes.bool,
    hasChanges: PropTypes.bool,
    stageId: PropTypes.number,
    onComplete: PropTypes.func,
    onCancel: PropTypes.func,
    draft: PropTypes.any.isRequired,
    updateStage: PropTypes.func.isRequired,
    updateDraft: PropTypes.func.isRequired,
  };

  static defaultProps = {
    show: false,
    draft: null,
    stageId: null,
    hasChanges: false,
    onComplete: () => {},
    onCancel: () => {},
  }

  onSave = () => {
    const stageId = this.props.stageId;

    this.props.updateStage(
      stageId,
      {
        skipLogic: this.props.draft,
      },
    );

    this.props.onComplete();
  };

  renderButtons() {
    return [].concat(
      this.props.hasChanges ? [<Button key="save" size="small" onClick={this.onSave}>Save</Button>] : [],
    );
  }

  render() {
    const {
      show,
      onCancel,
      updateDraft,
      draft: {
        filter,
        action,
        ...predicate
      },
    } = this.props;

    return (
      <ProtocolCard
        title="Edit skip logic"
        type="intent"
        buttons={this.renderButtons()}
        show={show}
        onCancel={onCancel}
      >
        <div className="edit-skip-logic">
          <div className="edit-skip-logic__section">
            <div className="edit-skip-logic__action">
              <DropDown
                options={toPairs({ SHOW: 'Show this stage if', SKIP: 'Skip this stage if' })}
                onChange={value => updateDraft({ action: value })}
                value={action}
              />
            </div>
          </div>
          <div className="edit-skip-logic__section">
            <div className="edit-skip-logic__rule">
              <NetworkRule
                logic={predicate}
                onChange={logic => updateDraft(logic)}
              />
            </div>
          </div>
          <div className="edit-skip-logic__section">
            <FilterGroup
              filter={filter}
              onChange={newFilter => updateDraft({ filter: newFilter })}
            />
          </div>
          <div className="edit-skip-logic__guidance">
            <div className="edit-skip-logic__bubble">
              Skip logic tells Network Canvas when to skip past a stage. Using it,
              you can create different pathways through your interview.
            </div>
          </div>
        </div>
      </ProtocolCard>
    );
  }
}

function getSkipLogic(stage) {
  if (!stage) { return null; }

  return (has(stage, 'skipLogic') ? stage.skipLogic : defaultLogic);
}

function makeMapStateToProps() {
  const getStage = makeGetStage();

  return function mapStateToProps(state, props) {
    const stage = getStage(state, props);
    const skipLogic = getSkipLogic(stage);

    return { draft: skipLogic };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    updateStage: bindActionCreators(stageActions.updateStage, dispatch),
  };
}

export { EditSkipLogic };

export default compose(
  connect(makeMapStateToProps, mapDispatchToProps),
  Draft,
)(EditSkipLogic);
