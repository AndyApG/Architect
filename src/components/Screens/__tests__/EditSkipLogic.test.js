/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { EditSkipLogic } from '../EditSkipLogic';

const mockProps = {
  draft: {},
  updateStage: () => {},
  updateDraft: () => {},
  rules: {},
  onChange: () => {},
};

describe('<EditForm />', () => {
  it('can render', () => {
    const component = shallow(<EditSkipLogic {...mockProps} />);

    expect(component).toMatchSnapshot();
  });
});
