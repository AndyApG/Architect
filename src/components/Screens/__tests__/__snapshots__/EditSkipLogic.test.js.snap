// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditForm /> can render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EditSkipLogic
    draft={Object {}}
    hasChanges={false}
    onChange={[Function]}
    onComplete={[Function]}
    rules={Object {}}
    show={false}
    stageId={null}
    state={null}
    updateDraft={[Function]}
    updateStage={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buttons": Array [],
      "children": <SkipLogicEditor
        onChange={[Function]}
        rules={Object {}}
      />,
      "onCancel": [Function],
      "show": false,
      "state": null,
      "type": "logic",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "onChange": [Function],
        "rules": Object {},
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buttons": Array [],
        "children": <SkipLogicEditor
          onChange={[Function]}
          rules={Object {}}
        />,
        "onCancel": [Function],
        "show": false,
        "state": null,
        "type": "logic",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onChange": [Function],
          "rules": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
